src/Lean/Server/FileWorker/SetupFile.lean

```diff
@@ -0,0 +1,128 @@
+/-
+Copyright (c) 2023 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Ullrich, Marc Huisinga
+-/
+import Init.System.IO
+import Lean.Server.Utils
+import Lean.Util.FileSetupInfo
+import Lean.Util.LakePath
+import Lean.LoadDynlib
+
+namespace Lean.Server.FileWorker
+
+open IO
+
+structure LakeSetupFileOutput where
+    spawnArgs : Process.SpawnArgs
+    exitCode  : UInt32
+    stdout    : String
+    stderr    : String
+
+partial def runLakeSetupFile
+    (m                 : DocumentMeta)
+    (lakePath filePath : System.FilePath)
+    (imports           : Array Import)
+    (handleStderr      : String → IO Unit)
+    : IO LakeSetupFileOutput := do
+  let mut args := #["setup-file", filePath.toString] ++ imports.map (toString ·.module)
+  if m.dependencyBuildMode matches .never then
+    args := args.push "--no-build"
+  let spawnArgs : Process.SpawnArgs := {
+    stdin  := Process.Stdio.null
+    stdout := Process.Stdio.piped
+    stderr := Process.Stdio.piped
+    cmd    := lakePath.toString
+    args
+  }
+  let lakeProc ← Process.spawn spawnArgs
+
+  let rec processStderr (acc : String) : IO String := do
+    let line ← lakeProc.stderr.getLine
+    if line == "" then
+      return acc
+    else
+      handleStderr line
+      processStderr (acc ++ line)
+  let stderr ← IO.asTask (processStderr "") Task.Priority.dedicated
+
+  let stdout := String.trim (← lakeProc.stdout.readToEnd)
+  let stderr ← IO.ofExcept stderr.get
+  let exitCode ← lakeProc.wait
+  return ⟨spawnArgs, exitCode, stdout, stderr⟩
+
+inductive FileSetupResultKind where
+  | success
+  | noLakefile
+  | importsOutOfDate
+  | error (msg : String)
+
+structure FileSetupResult where
+  kind          : FileSetupResultKind
+  srcSearchPath : SearchPath
+  fileOptions   : Options
+
+def FileSetupResult.ofSuccess (pkgSearchPath : SearchPath) (fileOptions : Options)
+    : IO FileSetupResult := do return {
+  kind          := FileSetupResultKind.success
+  srcSearchPath := ← initSrcSearchPath pkgSearchPath,
+  fileOptions
+}
+
+def FileSetupResult.ofNoLakefile : IO FileSetupResult := do return {
+  kind          := FileSetupResultKind.noLakefile
+  srcSearchPath := ← initSrcSearchPath
+  fileOptions   := Options.empty
+}
+
+def FileSetupResult.ofImportsOutOfDate : IO FileSetupResult := do return {
+  kind          := FileSetupResultKind.importsOutOfDate
+  srcSearchPath := ← initSrcSearchPath
+  fileOptions   := Options.empty
+}
+
+def FileSetupResult.ofError (msg : String) : IO FileSetupResult := do return {
+  kind          := FileSetupResultKind.error msg
+  srcSearchPath := ← initSrcSearchPath
+  fileOptions   := Options.empty
+}
```

Comment:
I would probably have compressed these into one function (`ofUnsuccessful (kind : FileSetupResultKind)`?) but it is definitely fine either way

Reply:
Good point!

Reply:
Ah, I think I actually prefer it the way it is now. `ofUnsuccessful FileSetupResultKind.successful` would be strange.

Reply:
Indeed it would be because for `successful` we should still use a separate function using the correct search path and options :) . Or do you mean the mere possibility of that call would be strange?

Reply:
The possibility for the call! The idea of this "design pattern" is that as long as you use the constructor functions that the type provides instead of the primitive constructor, you get a type that fulfills its invariants. `ofUnsuccessful` would violate this.